nixos/grub: light grub by default, 150MB closure size reduction

infinisil opened the pull request:
Since the default grub2 package has ZFS enabled and this was the one
that got used for NixOS and it's not being overridden, every NixOS user
gets the ZFS dependencies, even when they're not using ZFS at all.

This PR effectively removes 150MB from the system closure on a default minimal setup. Bringing the closure size for an empty `configuration.nix` from 1049MB to 903MB (tested with `nix path-info -S` of a `nix-build nixos -A config.system.build.toplevel`).

#### Can it break anything?
This switches `grub2` for `grub2_minimal` for NixOS. Since the latter is being defined like this https://github.com/NixOS/nixpkgs/blob/ad112ca0a48a25ba760a616a5eb18b85e0a82559/pkgs/top-level/all-packages.nix#L2833-L2835

and it's being overridden with `zfsSupport = true` when a ZFS filesystem is being used
https://github.com/NixOS/nixpkgs/blob/ad112ca0a48a25ba760a616a5eb18b85e0a82559/nixos/modules/system/boot/loader/grub/grub.nix#L12
(`grub2_light` for this PR), this can only break systems who relied on ZFS support using NixOS' non-fixed grub derivation propagated through here https://github.com/NixOS/nixpkgs/blob/ad112ca0a48a25ba760a616a5eb18b85e0a82559/nixos/modules/system/boot/loader/grub/grub.nix#L561-L567

Ping @aszlig @symphorien

###### Things done

I ran the NixOS test `installer.zfsroot` successfully. Also tried to run `installer.simpleUefiGrub`, but it failed for [some unrelated networking issue](https://gist.github.com/Infinisil/a20ff95ac0f7f436b178ae520f9a74f7).

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

infinisil committed:
nixos/grub: light grub by default

Since the default grub2 package has ZFS enabled and this was the one
that got used for NixOS and it's not being overridden, every NixOS user
gets the ZFS dependencies, even when they're not using ZFS at all.

Infinisil commented:
@aszlig Well spotted, I'll change that. Which test didn't succeed for you though? The networking issue I mentioned probably doesn't have anything to do with this

aszlig commented:
@Infinisil: I didn't run any tests, but it seems you did:

> I ran the NixOS test installer.zfsroot successfully. Also tried to run installer.simpleUefiGrub, but it failed for some unrelated networking issue.

aszlig committed:
nixos/tests/installer: Fix grub2 pkgs in VM deps

We don't have networking access within the test VMs, so with the change
to default to grub2_light we now no longer simply add pkgs.grub2 and
pkgs.grub2_efi to the closure, because the grub NixOS module overrides
it like this (simplified):

  grubReal = pkgs.grub2_light.override { zfsSupport = ...; };
  grubEfi = grubReal.override { efiSupport = ...; };

So if we'd simply use [ pkgs.grub2_light pkgs.grub2_efi ] we would end
up with the wrong derivation in EFI tests because pkgs.grub2_efi is
actually an override of grub2 *with* ZFS but the actual installer test
would use grub2 *without* ZFS and thus will try to rebuild grub.

Tested against all the installer NixOS tests and all of them still
succeed except the grub1 test (which already failed prior to the
previous change).

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
@Infinisil: Test fix added.

Infinisil commented:
Ah I see now, thanks a lot @aszlig

grahamc commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -9,12 +9,12 @@ let
 
   realGrub = if cfg.version == 1 then pkgs.grub
-    else if cfg.zfsSupport then pkgs.grub2.override { zfsSupport = true; }
+    else if cfg.zfsSupport then pkgs.grub2_light.override { zfsSupport = true; }
```
seems we should just be using regular grub in this case?

grahamc replied:
especially since:

```
2826  grub2_light = grub2.override {
2827    zfsSupport = false;
2828  };
```

Infinisil replied:
https://github.com/NixOS/nixpkgs/blob/129f94fb38787ef866805e0f38e4b5df8050c20c/pkgs/top-level/all-packages.nix#L2818-L2828

Infinisil replied:
The default grub has ZFS support enabled

Infinisil replied:
Oh I get what you mean, overriding light with zfs = true is the same as the normal one, gotcha

grahamc replied:
Yes, and that is what we want here:

 +    else if cfg.zfsSupport then pkgs.grub2_light.override { zfsSupport = true; }

aszlig replied:
Ideally the `grub2_light` variant should be the one used by the NixOS module, so that only the relevant features are enabled, because permuting all the features that might be available through the module as package definitions in `all-packages.nix` would be even more overkill.

Besides (and unrelated) these `grub2.override`s should really be `self.grub2.override` so they can be overridden via overlays.

grahamc commented on nixos/tests/installer.nix:
```diff
@@ -232,25 +233,29 @@ let
+            system.extraDependencies = let
+              # This is because the EFI version of grub overrides the base
+              # version of GRUB 2, so we can't just use pkgs.grub2_efi here.
+              grub2 = pkgs.grub2_light.override { zfsSupport = grubUseZfs; };
```
Should probably use if grubUseZfs then pkgs.grub2 else pkgs.grub2_light, and move the comment about the EFI stuff to below this line.

aszlig replied:
@grahamc: It's already a bit obscure that way, doing `if grubUseZfs then pkgs.grub2 else pkgs.grub2_light` IMHO makes it worse because then you have to look at the actual definition in `all-packages.nix` to check which flags are enabled (which itself again uses `.override`).

grahamc replied:
I guess my perspective is the `_light` variant becomes meaningless and overriding an overridden package to disable the override seems really weird.

aszlig replied:
I agree. Ideally the `grub2` package should be the main package and `grub2_full` should be an override of `grub2_light` with all that stuff enabled. In the line here `pkgs.grub2.override` would also work, because `grub2_full` is just `grub2_light` with ZFS support.

grahamc commented on nixos/tests/installer.nix:
```diff
@@ -232,25 +233,29 @@ let
+              # curl's tarball, we see what it's trying to download
+              curl
+            ] ++ optional (bootLoader == "grub" && grubVersion == 1) pkgs.grub
+              ++ optionals (bootLoader == "grub" && grubVersion == 2) [ grub2 grub2Efi ];
```
It feels a bit like much of this diff is moving code just to make the formatting slightly nicer. That can make it much more difficult to read the diff, for a minor improvement.

aszlig replied:
There is `git diff -w` which should make it way easier to follow.

aszlig replied:
(also GitHub has an option `Hide whitespace changes`, which does something similar)

grahamc commented:
Aside from some very minor changes, lgtm.

Infinisil commented:
So, should I mess around some more to try get this to look a bit better? Not sure if worth the effort even :)

matthewbauer reviewed (changes requested):
I am not a fan of using special cased things like this. I know they exist in all-packages.nix , but IMO they shouldn't be there.

The issue with something like this is that it will break overlays settings grub2. Previously you could make an overlay with something like:

```
self: super: {
  grub2 = super.grub2.override {
    zfsSupport = true;
    xenSupport = false;
    efiSupport = true;
  };
}
```

If you think we should use the grub2 light by default, you should make ```grub2``` in all-packages.nix point to it. Doing it this way is just kind of confusing.

Infinisil commented:
While this would be *very* nice, I really don't feel like doing this now. Maybe I'll do it later, but until then I'm closing this. Somebody else feel free to do this.
